mmengine/infer/infer.py

```diff
@@ -0,0 +1,474 @@
+
+    The BaseInferencer provide the standard workflow for inference as follows:
+
+    1. Preprocess the input data by :meth:`preprocess`.
+    2. Forward the data to the model by :meth:`forward`. ``BaseInferencer``
+       assumes the model inherits from :class:`mmengine.models.BaseModel` and
+       will call `model.test_step` in :meth:`forward` by default.
+    3. Visualize the results by :meth:`visualize`.
+    4. Postprocess and return the results by :meth:`postprocess`.
+
+    When we call the subclasses inherited from BaseInferencer(not override the
+    ``__call__``), the workflow will be executed in order.
+
+    All subclasses of BaseInferencer could define the following class
+    attributes for customization:
+
+    - ``preprocess_kwargs``: The keys of the kwargs that will be passed to
+      :meth:`preprocess`.
+    - ``forward_kwargs``: The keys of the kwargs that will be passed to
+      :meth:`forward`
+    - ``visualize_kwargs``: The keys of the kwargs that will be passed to
+      :meth:visualize
+    - ``postprocess_kwargs``: The keys of the kwargs that will be passed to
+      :meth:`postprocess`
+
+    All attributes mentioned above should be a ``set`` of keys(strings),
+    and each key should not be duplicated. Actually, :meth:`__call__` will
+    dispatch all the arguments to the corresponding methods according to the
+    ``xxx_kwargs`` mentioned above, therefore, the key in sets should
+    be unique to avoid ambiguous dispatching.
+
+    Warning:
+        If subclasses defined the class attributes mentioned above with
+        duplicated keys, an ``AssertionError`` will be raised during import
+        process.
+
+    Subclasses inherited from ``BaseInferencer`` should implement
+    :meth:`_init_pipeline`, :meth:`visualize` and :meth:`postprocess`:
+
+    - _init_pipeline: Return a callable object to preprocess the input data.
+    - visualize: Visualize the results returned by :meth:`forward`.
+    - postprocess: Postprocess the results returned by :meth:`forward` and
+      :meth:`visualzie`.
+
+    Args:
+        model (str, optional): Path to the config file or the model name
+            defined in metafile. Take the `mmdet metafile <https://github.com/open-mmlab/mmdetection/blob/master/configs/retinanet/metafile.yml>`_
+            as an example, the `model` could be `retinanet_r18_fpn_1x_coco` or
+            its alias.
+        weights: Path to the checkpoint. If it is not specified and model is a
+            model name of metafile, the weights will be loaded from metafile.
+            Defaults to None
+        device (str, optional): Device to run inference. If None, the best
+            device will be automatically used. Defaults to None.
+
+    Note:
+        Since `Inferencer` could be used to inference batch data, therefore
+        `collate_fn` should be defined. `collate_fn` is not defined in config
+        file, the `collate_fn` will be `pseudo_collate` by default.
```

Comment:
```suggestion
        Since `Inferencer` could be used to inference batch data, 
        `collate_fn` should be defined. If `collate_fn` is not defined in config
        file, the `collate_fn` will be `pseudo_collate` by default.
```